Feature: Implement configurable cursor shapes and color (reopened)

cypheon opened the pull request:
Re-opened this PR from #790

The change in `NvimView/NvimView/com.qvacua.NvimView.vim` is a first proposal, to restore the previous default cursor style for users without any `guicursor` config in their vimrc.

The `autocommand` for the cursor-styling highlight groups is necessary to prevent them being cleared on a colorscheme change. The naming should be specific enough that it does not accidentally collide with any user's custom highlight groups.
At first I was a bit hesitant to add an AutoCmd (for the highlight groups) to the default file, but then I noticed there are already lots of autocommands, even with an empty vimrc. So I hope it's okay.

This probably solves the currently open cursor-configuration related issues: #590, #758

qvacua commented on NvimView/NvimView/ModeInfo.swift:
```diff
@@ -0,0 +1,50 @@
+import MessagePack
+
+public enum CursorShape : Equatable {
+  case Block
```
For the sake of consistency, we should use lowercase for the `enum` cases.

cypheon replied:
Done.

qvacua commented on NvimView/NvimView/NvimView+Draw.swift:
```diff
@@ -96,14 +84,58 @@ extension NvimView {
+
+    // will be used for clipping
+    var cursorRect: CGRect
+    var cursorTextColor: Int
```
This can be `let`.

cypheon replied:
Done, converted `cursorTextColor` to `let`. As `cursorRect` is modified later (either width or height, depending on cursor shape), it's not easily possible there without making a temporary copy and then assigning that. (Which I think is not worth it.)

qvacua commented on NvimView/NvimView/com.qvacua.NvimView.vim:
```diff
@@ -1,2 +1,7 @@
+autocmd VimEnter,ColorScheme * :highlight default VimrDefaultCursor gui=reverse guibg=NONE guifg=NONE
+set guicursor=a:block-VimrDefaultCursor
+autocmd VimEnter,ColorScheme * :highlight default VimrInsertCursor guibg=fg
+set guicursor=i:ver25-VimrInsertCursor
```
I don't really understand what's going on here, but it seems to work 😀

cypheon replied:
This creates two "highlighting groups", "VimrDefaultCursor" and "VimrInsertCursor" (I hope those do not clash with anyone's custom vimrc). It then defines two cursor shapes, one for INSERT mode and one for all other modes, to use the newly created highlighting definitions.
Because custom highlighting groups are cleared on a colorscheme change, the ":highlight …" command is added as an autocmd, so the highlighting groups are recreated after every colorscheme change.

cypheon commented:
Thanks for the feedback! Sorry for the delay. I rebased my branch on the current `develop` branch and implemented the changes you suggested. Please let me know if there's anything left to be done before this can be merged :)
